Added import and export for slashing protection data.

rolfyone opened the pull request:
- `teku slashing-protection export --data-path=<PATH> --to=<FILE>`

 - `teku slashing-protection import --data-path=<PATH> --from=<FILE>`

 - `--data-path` defaults to the default path for teku.

 Exporting will only export to the minimal interchange format, as that is all that's stored by teku.

 On import, complete or minimal files are supported, and summarised for use by teku. If the genesis_validators_root does not match, the process will halt rather than corrupting existing data. If an existing file is found, the MAX of the existing data and imported data will be stored, to ensure there is no possibility of stepping 'backwards' in slashing protection data.

 fixes #2601

Signed-off-by: Paul Harris <paul.harris@consensys.net>

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

rolfyone committed:
Added import and export for slashing protection data.

 - teku slashing-protection export --data-path=<PATH> --to=<FILE>

 - teku slashing-protection import --data-path=<PATH> --from=<FILE>

 - --data-path defaults to the default path for teku.

 Exporting will only export to the minimal interchange format, as that is all that's stored by teku.

 On import, complete or minimal files are supported, and summarised for use by teku. If the genesis_validators_root does not match, the process will halt rather than corrupting existing data. If an existing file is found, the MAX of the exting data and imported data will be stored, to ensure there is no possibility of stepping 'backwards' in slashing protection data.

 fixes #2601

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
pretty print json

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
spotless

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
make NEVER_SIGNED be null, and conversion logic associated, as well as go back to using sync writer.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
typo

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,81 @@
+            "the genesisValidatorsRoot of "
+                + file.getName()
+                + " does not match the expected "
+                + genesisValidatorsRoot.toHexString().toLowerCase());
```
This should go via `SubcommandLogger.error`

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,81 @@
+  private void readSlashProtectionFile(final File file) {
+    try {
+      SlashingProtectionRecord slashingProtectionRecord =
+          yamlProvider.fileToObject(file, SlashingProtectionRecord.class);
```
I'm not convinced the `YamlProvider` should be dealing with file IO.  I'd expect it to mirror `JsonProvider` and just do the conversion to/from bytes (potentially to/from `InputStream` if we wanted to be able to stream the data.

rolfyone replied:
i dont mind the concept of passing input streams and output streams, but in terms of the provider dealing with the IO, i kind of think its cleaner... maybe not 100% the way it is now, but for the core processing to just get the data and use it, and this deals with reading/writing the yaml seems like a nice single responsibility...

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,81 @@
+                + genesisValidatorsRoot.toHexString().toLowerCase());
+        System.exit(1);
+      }
+      final String pubkey = file.getName().substring(0, file.getName().length() - 4);
```
```suggestion
      final String pubkey = file.getName().substring(0, file.getName().length() - ".yml".length());
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,81 @@
+      }
+      final String pubkey = file.getName().substring(0, file.getName().length() - 4);
+      minimalSigningHistoryList.add(
+          new MinimalSigningHistory(BLSPubKey.fromHexString(pubkey), slashingProtectionRecord));
```
Probably should catch the exception thrown when this filename turns out to not be a valid public key and provide a useful error message that includes the filename.

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,81 @@
+            new File(toFileName),
+            new MinimalSlashingProtectionInterchangeFormat(
+                new Metadata(InterchangeFormat.minimal, UInt64.valueOf(2L), genesisValidatorsRoot),
+                minimalSigningHistoryList));
```
Probably would be better to just convert to JSON with `jsonProvider.objectToPrettyJSON` then write to file rather than using `getObjectMapper` and having the Jackson specific code leak out to here.

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,137 @@
+    metadata = jsonMapper.treeToValue(jsonNode.get("metadata"), Metadata.class);
+    if (!metadata.interchangeFormatVersion.equals(UInt64.valueOf(2L))) {
+      System.err.println(
+          "Import file " + inputFile.toString() + " is not format version 2, cannot continue.");
```
`SubCommandLogger.error`

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,137 @@
+    }
+  }
+
+  private void summariseCompleteInterchangeFormat(
```
nit: I'd have this return the result rather than assign to `data` so the if above is symmetric with both versions assigning to data inline.

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,137 @@
+        new SlashingProtectionRecord(
+            lastSlot.orElse(UInt64.ZERO),
+            sourceEpoch.orElse(null),
+            targetEpoch.orElse(null),
```
```suggestion
            sourceEpoch.orElse(ValidatorSigningRecord.NEVER_SIGNED),
            targetEpoch.orElse(ValidatorSigningRecord.NEVER_SIGNED),
```

rolfyone replied:
Can't access that from this context, it'd be cyclic dependency...

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,137 @@
+  private void updateLocalRecord(final MinimalSigningHistory minimalSigningHistory) {
+    String validatorString = minimalSigningHistory.pubkey.toHexString().substring(2).toLowerCase();
+
+    System.out.println("Importing " + validatorString);
```
`SubCommandLogger.display`

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,137 @@
+      System.err.println(
+          "Validator "
+              + minimalSigningHistory.pubkey.toHexString()
+              + " has a different validators signing root to the data being imported");
```
I'd load all the existing validator records upfront and check they all match prior to starting to apply changes.  Otherwise we risk having some records imported and some not.

Basically I'd import by reading the incoming records, reading all the existing records, merging the two and writing them out.  It might still fail while writing out because of IO errors but at least you've done all the validation up-front and given yourself a chance.

Also `SubCommandLogger`

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/slashinginterchange/YamlProvider.java:
```diff
@@ -0,0 +1,88 @@
+
+  public <T> void writeToFile(File file, T object) throws IOException {
+    objectMapper.writerWithDefaultPrettyPrinter().writeValue(file, object);
+  }
```
Mentioned above, but I'd get rid of these or at most have versions that take `InputStream`/`OutputStream` so they don't care where the data is coming from.

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/signatures/SlashingProtector.java:
```diff
@@ -18,50 +18,64 @@
           final ValidatorSigningRecord signingRecord = loadSigningRecord(validator);
-          return handleResult(validator, signingRecord.maySignBlock(slot));
+          return handleResult(
+              validator, forkInfo.getGenesisValidatorsRoot(), signingRecord.maySignBlock(slot));
```
It would make more sense to pass genesis validators root into `maySignBlock` so that if it is allowed to sign, it just creates a new `ValidatorSigningRecord` that already holds the right genesis validators root.  It should also refuse to sign if it has a validators root that is different to what is passed in as that indicates that the slashing protection DB is from the wrong chain and something has gone wrong (make sure to log an error message in this case as well as the return or it won't be clear why it refused to sign).

rolfyone replied:
we discussed the validation against the existing validators root being a separate ticket? 

i've updated `maySignBlock` and `maySignAttestation` to have bytes32 instead of forkInfo come in

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/signatures/SlashingProtector.java:
```diff
@@ -74,15 +88,35 @@ private ValidatorSigningRecord loadSigningRecord(final BLSPublicKey validator)
+                  }
+                  return new SlashingProtectionRecord();
+                })
+            .map(ValidatorSigningRecord::fromSlashingProtectionRecord)
```
How did we manage to make this so much more complex?  Can we go back to how we were doing serialisation for `ValidatorSigningRecord`?  We seem to have wound up with yaml, IO and data model responsibilities scattered all through the code now.

rolfyone replied:
this is mostly because i had removed the separate yaml processor that had specific logic for how to export the validatorSigningRecord, and used the one in data. If we want two versions of yaml processor we could go back to the old version and clean that up a bit, but it was also doing things in a weird way... i could clean that up rather than pushing this through the other processor?

I guess i'd really like to have 1 yaml provider and 1 object definition for the serialized file, and its hard to access it from the import/export commands, so pushing it to one with minimal dependencies so that it could be reused was where i was headed...

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/signatures/record/ValidatorSigningRecord.java:
```diff
@@ -52,11 +60,24 @@ public ValidatorSigningRecord(
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
   }
```
Would be good to keep these as a simple delegation so that this class is focussed on being the data model and the details of serialisation are handled separately.

But they've also wound up only being used by tests which suggests the tests aren't actually testing what the production code now does.

rolfyone replied:
My original implementation had this all removed pretty much, and the tests actually wrote to file, and checked the file contents. It as closer to production because it was following it more faithfully, but we didn't want to remove the sync data writer, so i went back to this, but i think i preferred the old one more...

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/signatures/record/ValidatorSigningRecord.java:
```diff
@@ -92,24 +139,22 @@ public Bytes toBytes() {
   private boolean isSafeSourceEpoch(final UInt64 sourceEpoch) {
-    return attestationSourceEpoch.equals(NEVER_SIGNED)
-        || attestationSourceEpoch.compareTo(sourceEpoch) <= 0;
+    return attestationSourceEpoch == null || attestationSourceEpoch.compareTo(sourceEpoch) <= 0;
```
We should be consistent with using `NEVER_SIGNED` rather than using `null` directly.

ajsutton commented on ethereum/core/src/main/java/tech/pegasys/teku/core/signatures/record/ValidatorSigningRecord.java:
```diff
@@ -92,24 +139,22 @@ public Bytes toBytes() {
-  UInt64 getAttestationTargetEpoch() {
+  public UInt64 getAttestationTargetEpoch() {
     return attestationTargetEpoch;
   }
```
These shouldn't be public as we want to encapsulate the data as much as we can and just have the `maySign*` methods as the key public API.  They do appear to be unused now though.

ajsutton commented on teku/src/main/java/tech/pegasys/teku/cli/BeaconNodeCommand.java:
```diff
@@ -149,6 +152,9 @@
   @Mixin(name = "Data")
   private DataOptions dataOptions;
 
+  @Mixin(name = "DataStorage")
```
This is a human friendly name so:
```suggestion
  @Mixin(name = "Data Storage")
```

ajsutton commented on teku/src/main/java/tech/pegasys/teku/cli/slashingprotection/ExportCommand.java:
```diff
@@ -0,0 +1,80 @@
+
+@CommandLine.Command(
+    name = "export",
+    description = "Export slashing protection database in Minimal format.",
```
```suggestion
    description = "Export slashing protection database in minimal format.",
```

rolfyone replied:
you sure? minimal is the name of the format...

ajsutton replied:
Docs team might overrule me but even with it being the name of the format it's not a proper noun so doesn't deserve a capital.

ajsutton commented on teku/src/main/java/tech/pegasys/teku/cli/slashingprotection/ExportCommand.java:
```diff
@@ -0,0 +1,80 @@
+
+    SlashingProtectionExporter slashingProtectionExporter = new SlashingProtectionExporter();
+
+    SUB_COMMAND_LOG.display("Reading slashing protection data from: " + slashProtectionPath);
+    slashingProtectionExporter.initialise(slashProtectionPath);
```
We should probably display an error an exit if there are no slashing protection records found (ie the directory exists but is empty)

rolfyone replied:
I could potentially also output each file that I load? I ended up doing this on import, but maybe i should do it on export also...

ajsutton commented on teku/src/main/java/tech/pegasys/teku/cli/slashingprotection/ImportCommand.java:
```diff
@@ -0,0 +1,93 @@
+@CommandLine.Command(
+    name = "import",
+    description =
+        "Import slashing protection database. Supports Minimal or Complete interchange format.",
```
```suggestion
        "Import slashing protection database. Supports minimal or complete interchange format.",
```

ajsutton commented on teku/src/main/java/tech/pegasys/teku/services/ServiceController.java:
```diff
@@ -37,7 +39,9 @@ public ServiceController(final ServiceConfig config) {
-      services.add(ValidatorClientService.create(config));
+      final Path slashingProtectionPath =
+          DataOptions.getValidatorsSlashingProtectionPath(config.getConfig().getDataPath());
+      services.add(ValidatorClientService.create(config, slashingProtectionPath));
```
Calculating the slashingProtectionPath should stay inside `ValidatorClientService`.  The `ServiceController` class should know as little as possible about the details of the service and basically just create and pass the `ServiceConfig`.

rolfyone replied:
I had settled on this because DataOptions wasn't available to the validatorClient, and validatorClient wasn't available to the import/export, so I'll have to have a think about how this is all going to link together and come up with another solution i guess.

ajsutton replied:
Maybe just add a separate item to `TekuConfiguration` that has the slashing protection path.  Right now it would be calculated off of data dir but then we'd be in a good place to be able to make it a separate CLI arg in the future.

ajsutton reviewed (commented):
I think mostly we just need to tidy up how the serialisation is done.

Would have been way easier to review if this was done in a few smaller steps rather than all at once. :)

rolfyone committed:
review feedback

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
review feedback

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
merged 2601-signingrecord, and fixed build to work with the new signing record package.

also added some output and summaries to the command line output.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
updated interchange version to v.3

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
negative testing for import.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,96 @@
+  private void readSlashProtectionFile(final File file) {
+    try (InputStream input = Files.newInputStream(file.toPath())) {
+      ValidatorSigningRecord validatorSigningRecord =
+          ValidatorSigningRecord.fromBytes(Bytes.of(input.readAllBytes()));
```
nit: Since we're reading all the bytes at once anyway we can just use:
```suggestion
    try {
      ValidatorSigningRecord validatorSigningRecord =
          ValidatorSigningRecord.fromBytes(Bytes.of(Files.readAllBytes(file.toPath)));
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,96 @@
+          && !genesisValidatorsRoot.equals(validatorSigningRecord.getGenesisValidatorsRoot())) {
+        log.exit(
+            1,
+            "the genesisValidatorsRoot of "
```
```suggestion
            "The genesisValidatorsRoot of "
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java:
```diff
@@ -0,0 +1,96 @@
+            "the genesisValidatorsRoot of "
+                + file.getName()
+                + " does not match the expected "
+                + genesisValidatorsRoot.toHexString().toLowerCase());
```
Not sure why there's a `toLowerCase` on this.

rolfyone replied:
just to guarantee lower case, I can remove it... it was just something i got in the habit of when doing rest api because we always wanted it in lowercase...

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+                    jsonMapper.treeToValue(jsonNode.get("data"), CompleteSigningHistory[].class)));
+      }
+    } catch (JsonMappingException e) {
+      String cause = e.getCause() != null ? e.getCause().getMessage() : Strings.EMPTY;
```
```suggestion
      String cause = e.getCause() != null ? e.getCause().getMessage() : e.getMessage();
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+      String cause = e.getCause() != null ? e.getCause().getMessage() : Strings.EMPTY;
+      log.exit(1, "Failed to load data from " + inputFile.getName() + ". " + cause);
+    } catch (JsonParseException e) {
+      String cause = e.getCause() != null ? e.getCause().getMessage() : Strings.EMPTY;
```
```suggestion
      String cause = e.getCause() != null ? e.getCause().getMessage() : e.getMessage();
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+                + " does not appear to have metadata information, and cannot be loaded.");
+        return; // Testing mocks log.exit
+      }
+      if (!metadata.interchangeFormatVersion.equals(INTERCHANGE_VERSION)) {
```
```suggestion
      if (!INTERCHANGE_VERSION.equals(metadata.interchangeFormatVersion)) {
```

to avoid potential NullPointerException if the version is missing (Jackson might have caught it already but doesn't hurt to be paranoid)

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+                + inputFile.toString()
+                + " is not format version "
+                + INTERCHANGE_VERSION.toString()
+                + ", cannot continue.");
```
```suggestion
            "Import file "
                + inputFile.toString()
                + " has unsupported format version "
                + metadata.interchangeFormatVersion
                + ". Required version is "
                + INTERCHANGE_VERSION);
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+  }
+
+  private void updateLocalRecord(final MinimalSigningHistory minimalSigningHistory) {
+    String validatorString = minimalSigningHistory.pubkey.toHexString().substring(2).toLowerCase();
```
```suggestion
    String validatorString = minimalSigningHistory.pubkey.toUnprefixedHexString().toLowerCase();
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+    String validatorString = minimalSigningHistory.pubkey.toHexString().substring(2).toLowerCase();
+
+    log.display("Importing " + validatorString);
+    Path outputFile = slashingProtectionPath.resolve(validatorString.concat(".yml"));
```
```suggestion
    Path outputFile = slashingProtectionPath.resolve(validatorString + ".yml");
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+    if (outputFile.toFile().exists()) {
+      try (InputStream input = Files.newInputStream(outputFile)) {
+        existingRecord =
+            Optional.ofNullable(ValidatorSigningRecord.fromBytes(Bytes.of(input.readAllBytes())));
```
```suggestion
      try {
        existingRecord =
            Optional.ofNullable(ValidatorSigningRecord.fromBytes(Bytes.of(Files.readAllBytes(outputFile)));
```

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java:
```diff
@@ -0,0 +1,171 @@
+              .toValidatorSigningRecord(existingRecord, metadata.genesisValidatorsRoot)
+              .toBytes()
+              .toArray());
+    } catch (IOException e) {
```
Actually probably should use `SyncDataAccessor` here to make sure the change is really persisted to disk.  Likely have to move it to a different module but it's very standalone so that should be easy enough.

ajsutton commented on data/dataexchange/src/main/java/tech/pegasys/teku/data/slashinginterchange/MinimalSigningHistory.java:
```diff
@@ -0,0 +1,105 @@
+  @JsonProperty("pubkey")
+  public final BLSPubKey pubkey;
+
+  @JsonProperty("last_signed_block_slot")
+  public final UInt64 lastSignedBlockSlot;
+
+  @JsonProperty("last_signed_attestation_source_epoch")
+  public final UInt64 lastSignedAttestationSourceEpoch;
+
+  @JsonProperty("last_signed_attestation_target_epoch")
```
I think these are going to be renamed - probably wind up with a version 4.  Ok to merge as-is but we probably should wait for that change to come through before we really announce this so that we don't have to maintain support for version 3.

rolfyone replied:
interesting, but sounds like not a merge issue, so will mark resolved

ajsutton reviewed (commented):
Looking good, pretty much just little details to tweak now.

rolfyone committed:
Update data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

rolfyone committed:
Update data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

rolfyone committed:
Update data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

rolfyone committed:
Update data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionExporter.java

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

rolfyone committed:
Update data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

rolfyone committed:
Update data/dataexchange/src/main/java/tech/pegasys/teku/data/SlashingProtectionImporter.java

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

rolfyone committed:
review feedback

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch '2601' of https://github.com/rolfyone/artemis into 2601

rolfyone committed:
use SyncDataAccessor in import/export operations.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

ajsutton commented on data/files/src/main/java/tech/pegasys/teku/data/files/SyncDataAccessor.java:
```diff
@@ -11,7 +11,7 @@
  */
 
-package tech.pegasys.teku.core.signatures;
+package tech.pegasys.teku.data.files;
```
Since we're creating a new module (which is a good idea), I'd call it `:infrastructure:io` rather than `:data:files`.

ajsutton commented on teku/src/main/java/tech/pegasys/teku/services/ServiceController.java:
```diff
@@ -37,7 +38,8 @@ public ServiceController(final ServiceConfig config) {
     } else {
-      services.add(ValidatorClientService.create(config));
+      final Path slashingProtectionPath = config.getConfig().getValidatorsSlashingProtectionPath();
+      services.add(ValidatorClientService.create(config, slashingProtectionPath));
```
We can just pass in `config` now and have `ValidatorClientService` get the slashing protection path from the config itself.

rolfyone committed:
rename new package for SyncDataAccessor to infrastructure:io

 - also get slashingProtection path from tekuConfig rather than passing into ValidatorClientService.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into 2601

rolfyone committed:
Apply suggestions from code review

Co-authored-by: Adrian Sutton <adrian@symphonious.net>

ajsutton reviewed (approved):
LGTM.

rolfyone committed:
fix build

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
unit test fix

Signed-off-by: Paul Harris <paul.harris@consensys.net>
